build.sh

```diff
@@ -21,24 +23,78 @@ OL_UBI_IMAGE="${OL_UBI_IMAGE:-openliberty/open-liberty:20.0.0.10-kernel-java11-o
 #
 # The name and tag of the "stack image you will build.  This will used to create your inner loop development containers, and also as the base image for the first stage of your outer loop image build.
 #
-STACK_IMAGE="${STACK_IMAGE:-openliberty/application-stack:0.3}"
+STACK_IMAGE="${STACK_IMAGE:-openliberty/application-stack:0.4}"
 
 #
 # URL at which your outer loop Dockerfile is hosted
 #
-DEVFILE_DOCKERFILE_LOC="${DEVFILE_DOCKERFILE_LOC:-https://raw.githubusercontent.com/OpenLiberty/application-stack/master/outer-loop/0.3/Dockerfile}"
+DEVFILE_DOCKERFILE_LOC="${DEVFILE_DOCKERFILE_LOC:-https://raw.githubusercontent.com/OpenLiberty/application-stack-registry/releases/outer-loop/0.4/Dockerfile}"
 
 #
 # URL at which your outer loop deploy YAML template is hosted
 #
-DEVFILE_DEPLOY_YAML_LOC="${DEVFILE_DEPLOY_YAML_LOC:-https://raw.githubusercontent.com/OpenLiberty/application-stack/master/outer-loop/0.3/app-deploy.yaml}"
+DEVFILE_DEPLOY_YAML_LOC="${DEVFILE_DEPLOY_YAML_LOC:-https://raw.githubusercontent.com/OpenLiberty/application-stack-registry/releases/outer-loop/0.4/app-deploy.yaml}"
+
+generate() {
+
+    # Base customization.
+    mkdir -p generated
+    sed -e "s!{{.OL_RUNTIME_VERSION}}!$OL_RUNTIME_VERSION!; s!{{.STACK_IMAGE}}!$STACK_IMAGE!; s!{{.DEVFILE_DOCKERFILE_LOC}}!$DEVFILE_DOCKERFILE_LOC!; s!{{.DEVFILE_DEPLOY_YAML_LOC}}!$DEVFILE_DEPLOY_YAML_LOC!" templates/devfile.yaml > generated/devfile.yaml
+    sed -e "s!{{.BASE_OS_IMAGE}}!$BASE_OS_IMAGE!; s!{{.OL_RUNTIME_VERSION}}!$OL_RUNTIME_VERSION!" templates/stackimage/Dockerfile > generated/stackimage-Dockerfile
+
+    # Outer loop customization of Dockerfile
+    sed -e "s!{{.STACK_IMAGE}}!$STACK_IMAGE!; s!{{.OL_UBI_IMAGE}}!$OL_UBI_IMAGE!" templates/outer-loop/Dockerfile > generated/Dockerfile
+
+}
+
+release() {
+
+    SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+    REGISTRY_HOME=$SCRIPT_DIR/../application-stack-registry
+    RELEASES_DIR=$REGISTRY_HOME/releases
+
+    # copy devfile to releases
+    DIRS="outer-loop/0.4 outer-loop/0.4.0 outer-loop/latest"
+    for d in $DIRS; do 
+        mkdir -p $RELEASES_DIR/$d 
+        cp generated/Dockerfile $RELEASES_DIR/$d 
+        # (no customization at present)
+        cp templates/outer-loop/app-deploy.yaml $RELEASES_DIR/$d
+    done
+
+    # copy devfile to releases
+    DIRS="devfile/0.4 devfile/0.4.0  devfile/latest"
+    mkdir -p $DIRS
+    for d in $DIRS; do 
+        mkdir -p $RELEASES_DIR/$d 
+        cp generated/devfile.yaml $RELEASES_DIR/$d
+        cp generated/meta.yaml $RELEASES_DIR/$d
+    done
+
+    # copy devfile to devfiles (for 'odo create')
+    cp generated/devfile.yaml $REGISTRY_HOME/devfiles/java-openliberty
```

Comment:
Should you assume that $REGISTRY_HOME/devfiles/java-openliberty exists?

Reply:
Added check